src/Lti/Factory/LtiRequestFactory.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+/**
+ *  This program is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU General Public License
+ *  as published by the Free Software Foundation; under version 2
+ *  of the License (non-upgradable).
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ *  Copyright (c) 2020 (original work) Open Assessment Technologies S.A.
+ */
+
+declare(strict_types=1);
+
+namespace OAT\SimpleRoster\Lti\Factory;
+
+use OAT\SimpleRoster\Lti\Exception\InvalidLtiVersionException;
+use OAT\SimpleRoster\Lti\Request\LtiRequest;
+
+class LtiRequestFactory
+{
+    /** @var string */
+    private $ltiVersion;
+
+    /** @var Lti1p1RequestFactory */
+    private $lti1p1RequestFactory;
+
+    /** @var Lti1p3RequestFactory */
+    private $lti1p3RequestFactory;
+
+    public function __construct(
+        string $ltiVersion,
+        Lti1p1RequestFactory $lti1p1RequestFactory,
+        Lti1p3RequestFactory $lti1p3RequestFactory
+    ) {
+        $this->ltiVersion = $ltiVersion;
+        $this->lti1p1RequestFactory = $lti1p1RequestFactory;
+        $this->lti1p3RequestFactory = $lti1p3RequestFactory;
+    }
```

Comment:
Just not sure if it can't be done the other way, but why this "factory" should have both services injected in any case, not depending on which is actually used? Shouldn't it create them depending on env variable?

Reply:
Yes, this is conceptually wrong. Basically in any case you are instantiating one of the factories for no reason. They must be instantiated here. Please check for example here: OAT\SimpleRoster\Lti\LoadBalancer\LtiInstanceLoadBalancerFactory`

Reply:
If I change it, I will need to add all the parameter from both factories on the constructor here. Are you sure I need to change it?

Reply:
IMO - yes.
Let me explain a bit:
IMO - it would be better to have this as a factory class, which indeed is injected by all the args/services needed to build "versioned factories" (BTW, since they're not real factories, I would rename them as `Lti1p1RequestBuilder` / `Lti1p3RequestBuilder`). Even if it will increase the class complexity (which we will be trying to deal with for other services in next sprint), it would be more "cleaner" way as for me

Reply:
you can actually put the suggested builders (`Lti1p1RequestBuilder` and `Lti1p3RequestBuilder`) behind a `LtiRequestBuilderInterface`, and use `SF's service factories` to return the wanted instance based on your app configuration (will be autoresolved and autowired where needed).

imo, will ease maintenance, you can focus on business logic. 

see https://symfony.com/doc/current/service_container/factories.html#passing-arguments-to-the-factory-method